File: sdks/rust/examples/basic.rs
//! Basic reconpro-sdk usage: version ping + typed scan result.
//!
//! Run:  cargo run --example basic [-- <target>]
//!
//! By default it scans a nonexistent "*.invalid" target — fully offline
//! (the CLI truth layer short-circuits unreachable targets instantly).
//! A custom target:  cargo run --example basic -- example.com

fn main() {
    let target = std::env::args().nth(1).unwrap_or_else(|| "nonexistent-demo.invalid".into());
    let client = reconpro_sdk::Client::new();

    match client.version() {
        Ok(version) => println!("version         : {version}"),
        Err(e) => {
            eprintln!("version error: {e}");
            std::process::exit(1);
        }
    }

    match client.scan(&target) {
        Ok(result) => {
            println!("target          : {}", result.target);
            println!("grade           : {} (score {})", result.grade, result.total_score);
            println!("total findings  : {}", result.total_findings);
            println!("severity counts : {:?}", result.severity_counts);
            if let Some(tv) = &result.target_validation {
                println!(
                    "target state    : {} (dns={} reachable={} http_ok={})",
                    tv.state,
                    tv.dns_resolved.map(|b| b.to_string()).unwrap_or_else(|| "unknown".into()),
                    tv.reachable.map(|b| b.to_string()).unwrap_or_else(|| "unknown".into()),
                    tv.http_ok.map(|b| b.to_string()).unwrap_or_else(|| "unknown".into()),
                );
            }
            if let Some(meta) = &result.scan_metadata {
                println!("scan            : {} in {}s", meta.result.as_deref().unwrap_or("-"), meta.duration_s.unwrap_or(0.0));
            }
            for f in result.findings.iter().take(3) {
                println!("  - [{}] {}", f.severity, f.title);
                println!(
                    "    confidence={} verification_state={}",
                    f.confidence.map(|c| format!("{c:.2}")).unwrap_or_else(|| "n/a".into()),
                    f.verification_state.as_deref().unwrap_or("-"),
                );
            }
            if result.unreachable() {
                println!("truth layer     : no claims made about an unreachable target");
            }
        }
        Err(e) => {
            eprintln!("scan error: {e}");
            std::process::exit(1);
        }
    }
}

File: sdks/typescript/examples/basic.ts
/**
 * Basic reconpro-sdk usage: version ping + typed scan result + export.
 *
 * Run:  bun run examples/basic.ts [target]
 *
 * Binary resolution: `RECONPRO_BINARY` env → "reconpro" on PATH. For a fully
 * offline demo without the real CLI installed, point it at the test stub:
 *
 *   RECONPRO_BINARY=tests/fixtures/reconpro-stub.mjs bun run examples/basic.ts
 *
 * By default it scans a nonexistent `*.invalid` target — fully offline (the
 * CLI truth layer short-circuits unreachable targets instantly), so this
 * example needs no network access. Pass a real target for a live scan.
 */

import {
  ReconProClient,
  ReconProError,
  UNREACHABLE_TARGET,
} from "../src/index.js";

const target = process.argv[2] ?? "nonexistent-demo.invalid";
const client = new ReconProClient();

try {
  console.log(`version         : ${await client.version()}`);

  const result = await client.scan(target);
  console.log(`target          : ${result.target}`);
  console.log(`grade           : ${result.grade} (score ${result.totalScore})`);
  console.log(`total findings  : ${result.totalFindings}`);
  console.log(`severity counts : ${JSON.stringify(result.severityCounts)}`);
  if (result.targetValidation !== null) {
    const tv = result.targetValidation;
    console.log(
      `target state    : ${tv.state} (dns=${tv.dnsResolved} reachable=${tv.reachable} http_ok=${tv.httpOk})`,
    );
  }
  if (result.scanMetadata !== null) {
    console.log(
      `scan            : ${result.scanMetadata.result} in ${result.scanMetadata.durationS}s`,
    );
  }
  for (const f of result.findings.slice(0, 3)) {
    console.log(`  - [${f.severity}] ${f.title}`);
    console.log(
      `    confidence=${f.confidence} verification_state=${f.verificationState}`,
    );
  }
  if (result.unreachable) {
    console.log(
      `truth layer     : no claims made about an unreachable target (state=${UNREACHABLE_TARGET})`,
    );
  }

  // Export the last scan (format from the extension; default path in tmpdir).
  const report = await client.exportReport("sarif");
  console.log(`export          : ${report}`);

  // Unknown/future JSON keys stay reachable via `raw`.
  console.log(
    `raw keys sample : ${Object.keys(result.raw).slice(0, 4).join(", ")}...`,
  );
} catch (err) {
  if (err instanceof ReconProError) {
    console.error(
      `sdk error [${err.code}] exit=${err.exitCode}: ${err.message}`,
    );
    if (err.stderr) console.error(`stderr: ${err.stderr.slice(0, 300)}`);
    process.exit(1);
  }
  throw err;
}
